source/engines/rollEngine.js

```diff
@@ -1,548 +1,771 @@
-		if (acc[0] == '║' || cur == '') {
-			return acc;
-		} else if (/([0-9]+|[0-9]+d[0-9]+([dk][lh]?[0-9]+)?)([\+\-\*]([0-9]+|[0-9]+d[0-9]+([dk][lh]?[0-9]+)?))*$/.test(cur) || /[0-9]+$/.test(cur) || /\)/.test(cur)) {
-			acc.unshift('║');
-			return acc;
-		} else {
-			acc.unshift(cur);
-			return acc;
-		}
-	}, []);
-	textArray[0] == '║' ? textArray.shift() : ()=>{};
-	return new ResultBundle(rollArray.join(''),textArray.join(' '));
-}
-
-function getRollString(input, frac = false, simple = false) {
-	return parseRoll(input).toString(frac, simple);
-}
-
-module.exports = {parseRoll, getRollString}
+class Result {
+	#result;
+
+	constructor(result) {
+		this.#result = result;
+	}
+
+	getResult() {
+		return this.#result;
+	}
+
+	toString() {
+		return `${this.#result}`;
+	}
+}
+
+class RollResult extends Result {
+	#dieSides;
+
+	constructor(dieSides) {
+		super(Math.floor(Math.random() * dieSides) + 1);
+		this.#dieSides = dieSides;
+	}
+
+	getMaxValue() {
+		return this.#dieSides;
+	}
+
+	toString(frac = false) {
+		return frac ? `${this.getResult()}/${this.getMaxValue()}` : `${this.getResult()}`;
+	}
+}
+
+class StringResult extends Result {
+	#optList;
+
+	constructor(result, optList) {
+		super(result);
+		this.#optList = optList;
+	}
+
+	getMaxValue() {
+		return this.getResult();
+	}
+
+	toString(frac = false) {
+		return frac ? `${this.getResult()} from the list ${this.#optList}` : this.getResult();
+	}
+}
+
+const two_ops = { // Mathematical operations we currently support for rolls that use two arguments
+	ADD: '+',
+	SUBTRACT: '-',
+	MULTIPLY: '*'
+}
+
+const keep_ops = { // Operations for keeping or dropping rolls of a result set
+	KEEP: 'k',
+	KEEP_HIGHEST: 'kh',
+	DROP_LOWEST: 'dl',
+	DROP: 'd',
+	DROP_HIGHEST: 'dh',
+	KEEP_LOWEST: 'kl'
+}
+
+class ResultSet {
+
+	constructor() {}
+
+	static #surroundedByParens(str, symbolPos) {
+		var rightParensUnenclosed = 0,
+			leftParensUnenclosed = 0;
+		var leftSide = str.slice(0,symbolPos),
+			rightSide = str.slice(symbolPos+1,str.length-1);
+		for (var i = leftSide.length - 1; i >= 0 ; i-=1) {
+			if (leftSide[i] == ')' && leftParensUnenclosed <= 0) {
+				leftParensUnenclosed -= 1;
+			} else if (leftSide[i] == '(') {
+				leftParensUnenclosed += 1;
+			}
+		}
+		for (var i = 0; i < rightSide.length; i+=1) {
+			if (rightSide[i] == '(' && rightParensUnenclosed <= 0) {
+				rightParensUnenclosed -= 1;
+			} else if (rightSide[i] == ')') {
+				rightParensUnenclosed += 1;
+			}
+		}
+		return rightParensUnenclosed > 0 && leftParensUnenclosed > 0;
+	}
+
+	static #parseStringList(str) {
+		let strList = [];
+		let newEntry = "";
+		let startQuote = false;
+		let endQuote = false;
+		let spaceCount = 0;
+		for (var i = 0; i < str.length; i++) {
+			if (
+				(str[i] == '[' && i == 0) ||
+				(!(startQuote && endQuote) && str[i] == ']' && i == str.length-1) ||
+				(startQuote && endQuote && str[i] == ' ')
+				) {
+					if (str[i] == ' ') {
+						spaceCount++;
+					}
+				continue;

```

Comment:
I think incrementing `spaceCount` is enough of a difference in behavior that the `str[i] == '['` case and the `str[i] == ' '` case should be separated.

Reply:
I'm...inclined to agree. I think I added this at least in part because I ran into an issue with this during testing.